system/lib/wasmfs/js_api.cpp

```diff
@@ -188,4 +190,44 @@ void _wasmfs_readdir_finish(struct wasmfs_readdir_state* state) {
   free(state);
 }
 
+// Given a path, parse in into a full absolute path relative to the current
+// directory. Returns an allocated buffer that the caller must free, or 0 on
+// error.
+char* _wasmfs_parse_path(char* path) {
+  auto parsed = path::parseFile((char*)path);
+  if (auto err = parsed.getError()) {
+    return 0;
+  }
+  auto& file = parsed.getFile();
+
+  // Traverse up and build the full path.
+  std::vector<std::string> chunks;
+  while (1) {
+    auto parent = file->locked().getParent();
+    if (parent == file) {
+      break;
+    }
+    chunks.push_back(parent->locked().getName(file));
+    file = parent;
+  }
+
+  std::string ret;
+  for (auto it = chunks.rbegin(); it != chunks.rend(); ++it) {
+    auto& chunk = *it;
+    if (chunk.empty()) {
+      continue;
+    }
+    ret += '/';
+    ret += *it;
+  }
+  // The caller must free this buffer.
+  char* buffer = (char*)malloc(ret.size() + 1);
+  if (!buffer) {
+    return 0;
+  }
+  std::copy(ret.begin(), ret.end(), buffer);
+  buffer[ret.size()] = 0;
```

Comment:
How about just `return strdpy(ret.c_str())` instead of these last 7 lines?

Reply:
I was trying to use a more C++ fashion, I guess :smile: but sgtm